linux/libdriver/examples/rollup.c

```diff
@@ -0,0 +1,37 @@
+#include<stdio.h>
+#include<stdlib.h>
+#include<string.h>
+#include"rollup.h"
+
+int main(int argc, char *argv[])
+{
+	rollup_driver_t R[1];
+	uint64_t n, rxmax, txmax;
+	uint8_t hash[32];
+
+	int rc = 0;
+	if ((rc = rollup_driver_init(R)))
+		return EXIT_FAILURE;
+
+	const void *rx = rollup_driver_get_rx(R, &rxmax);
+	void       *tx = rollup_driver_get_tx(R, &txmax);
+
+	for (;;) {
+		int type = rollup_driver_loop(R, hash, &n);
```

Comment:
I think this name is a bit weird. You are accepting the previous input and requesting the next, atomically. Did you mean "iteration"? Even then it I believe it is a bit weird. Especially because there is a `reject_input` API function as well. So the default is to accept implicitly if you don't reject explicitly? You certainly have to call `loop` after calling reject input, right? To get the next one. But then you don't need a hash. You also don't need a hash if it was an inspect. So this is a bit confusing.

Reply:
I think the user has to write the hash to tx before caling `loop` in case he wants to accept. Perhaps `loop` should be called `get_next_request` and it should receive a constant that can be `ACCEPT_PREVIOUS` or `REJECT_PREVIOUS`.

Reply:
> I think this name is a bit weird. You are accepting the previous input and requesting the next, atomically. Did you mean "iteration"?

It is hard to find a name that conveys all the concepts `rollup_driver_loop` does. It accepts input[i-1], gets the length and type for input[i] but also marks the "checkpoint" in case we reject input[i]. I'm fine with `get_next_request`, but that still doesn't convey all of this, same as `finish` also didn't.

> You certainly have to call loop after calling reject input, right? To get the next one.

You don't. your program got reverted back to the last `loop` call you did, as well as all your state. The driver will resume from there and fill in all the values correctly. Think `[[noreturn]]`.
The idea is for `WAIT_FOR_INPUT` and `REJECT_INPUT` to mimic the `setjmp` / `longjmp` semantic in the control flow, it makes error handling easier. You can reject from nested calls for instance.

Renaming `rollup_driver_reject_input` to `rollup_driver_revert` may be clearer.

> You also don't need a hash if it was an inspect. So this is a bit confusing.

I can specify that `hash` gets ignored on inspects on the docs. As for doing in the API itself I'm not sure what to about it. Having a separate call for advance and inspect feels just worse IMO.

This is true, you'll also get reverted without rejecting the input on inspects, complexity of the

Reply:
I'm also considering getting the rx buffer from:
`rollup_driver_loop(R, hash, &n);` -> `rollup_driver_loop(R, hash, &n, &rx);`,
so that people don't try to pass the size of `hash` in.